text/0000-argument-spread-splarguments.md

```diff
@@ -0,0 +1,81 @@
+- Start Date: 2019-06-05
+- Relevant Team(s): Ember.js
+- RFC PR: (after opening the RFC PR, update this with a link to it and update the file name)
+- Tracking: (leave this empty)
+
+# Forwarding Component Arguments with "Splarguments"
+
+## Summary
+
+Currently it is possible to forward element attributes and modifiers(?) into child components for use as "splattributes" 
+such as, `...attributes`. This RFC proposes adding similar syntax and functionality for arguments such as, `...arguments`. 
+
+## Motivation
+
+The reason for doing this is to allow less verbose component templates where the verbosity only makes the template harder to read.
+Having the ability to spread arguments would make reading deeply nested component invocations much cleaner and nicer.
+
+
+For example, given the following invocation:
+```hbs
+<Foo @foo="bar" />
+```
+...and the following layout for the `Foo` component:
+```hbs
+<Bar ...arguments />
+```
+... and the following layout for the `Bar` component:
+```hbs
+<div>@foo</div>
+```
+Ember will render the folliwing HTML content:
+```html
+<div>bar</div>
+```
+
+A possible example in the wild could be 
+[ember-power-select](https://github.com/cibernox/ember-power-select/blob/master/addon/templates/components/power-select.hbs).
+This component has to explicitely pass on many arguments to its child components. With "splarguments" half or more of this template could
+be cut down.
```

Comment:
So I think a very important point to make here is that this also has a knock-on effect which reduces _maintenance_ of components. For instance, currently if I want to wrap a component, I have to explicitly forward every argument:

```hbs
<Foo @arg1={{@arg1}} @arg2={{@arg2}} />
```

This is already super verbose and pretty bad, but now let's say that I add `@arg3` to `Foo` - that means that I have to go through and update _every_ single component template that is wrapping `Foo`.

I think a great example of this would be a `Button` component, with a bunch of wrapping button "types":

```hbs
<!-- success-button.hbs -->
<Button @type="success" />
```


```hbs
<!-- warning-button.hbs -->
<Button @type="warning" />
```


```hbs
<!-- error-button.hbs -->
<Button @type="error" />
```

This would be a great pattern! No need to subclass the `Button` component, you can just have a bunch of TO-components that wrap it and provide the `@type` - until you need to start adding other arguments, like `@onClick` and so on. Now you have 3 files to go update, which is a giant pain.

Reply:
Updated to include these examples. What do you think?